File: voicecal-nlu/src/main/java/com/voicecal/nlu/time/TimeParserService.java
package com.voicecal.nlu.time;

import jakarta.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.lang.reflect.Method;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Service for parsing Chinese natural language time expressions into LocalDateTime objects.
 * Integrates the Time-NLP library (com.time.nlp) with fallback regex-based parsing.
 *
 * Supported patterns include:
 * - Relative dates: "明天", "后天", "三天后", "大后天"
 * - Fuzzy times: "下午", "傍晚", "早上", "中午"
 * - Compound expressions: "明天下午两点", "下周一上午十点"
 * - Absolute dates: "2024年3月15日", "3月15号下午3点"
 * - Day-of-week references: "下周三", "上周五"
 */
@Slf4j
@Service
public class TimeParserService {

    private static final DateTimeFormatter NORM_FORMATTER =
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private static final DateTimeFormatter NORM_DATE_FORMATTER =
            DateTimeFormatter.ofPattern("yyyy-MM-dd");

    // Time-NLP library accessed via reflection (not on Maven Central)
    private Object timeNormalizer;
    private Method parseMethod;
    private Method getTimeUnitMethod;
    private Method getTimeNormMethod;
    private Method getIsAllDayTimeMethod;
    private Method getOriginMethod;

    // ========== Fallback regex patterns for common Chinese time expressions ==========

    /** Matches relative day references: 明天, 后天, 大后天, 昨天, 前天 */
    private static final Pattern RELATIVE_DAY_PATTERN = Pattern.compile(
            "(大后天|后天|明天|昨天|前天|今天|今晚)"
    );

    /** Matches "N天后" / "N天以后" style relative offsets (Arabic or Chinese numerals) */
    private static final Pattern DAYS_LATER_PATTERN = Pattern.compile(
            "([一二两三四五六七八九十\\d]+)\\s*天(?:以后|后)"
    );

    /** Matches "N个小时后" / "N小时后" */
    private static final Pattern HOURS_LATER_PATTERN = Pattern.compile(
            "(\\d+)\\s*(?:个)?小时(?:以后|后)"
    );

    /** Matches "N分钟后" / "N分钟以后" */
    private static final Pattern MINUTES_LATER_PATTERN = Pattern.compile(
            "(\\d+)\\s*分钟(?:以后|后)"
    );

    /** Matches week references: 下周一, 上周三, 本周五 */
    private static final Pattern WEEK_REF_PATTERN = Pattern.compile(
            "(上|下|本)(?:周|星期|礼拜)([一二三四五六日天])"
    );

    /** Matches fuzzy time-of-day: 早上, 上午, 中午, 下午, 傍晚, 晚上, 凌晨 */
    private static final Pattern TIME_OF_DAY_PATTERN = Pattern.compile(
            "(凌晨|早上|上午|中午|下午|傍晚|晚上|晚间)"
    );

    /** Matches explicit hour: "两点", "十点", "3点", "14点", "两点半", "十点十五分" */
    private static final Pattern HOUR_PATTERN = Pattern.compile(
            "(\\d|[一二两三四五六七八九十]+)\\s*点(?:(半|[一二三四五六七八九十\\d]+)\\s*分?)?"
    );

    /** Matches absolute dates: 2024年3月15日, 3月15号 */
    private static final Pattern ABSOLUTE_DATE_PATTERN = Pattern.compile(
            "(?:(\\d{2,4})\\s*年)?(\\d{1,2})\\s*月(\\d{1,2})\\s*[日号]"
    );

    /** Chinese numeral to integer mapping */
    private static final java.util.Map<String, Integer> CHINESE_NUM_MAP;
    static {
        java.util.Map<String, Integer> m = new java.util.HashMap<>();
        m.put("零", 0); m.put("一", 1); m.put("二", 2); m.put("两", 2);
        m.put("三", 3); m.put("四", 4); m.put("五", 5); m.put("六", 6);
        m.put("七", 7); m.put("八", 8); m.put("九", 9); m.put("十", 10);
        m.put("十一", 11); m.put("十二", 12); m.put("十三", 13);
        m.put("十四", 14); m.put("十五", 15); m.put("十六", 16);
        m.put("十七", 17); m.put("十八", 18); m.put("十九", 19);
        m.put("二十", 20); m.put("二十一", 21); m.put("二十二", 22);
        m.put("二十三", 23); m.put("二十四", 24);
        // half
        m.put("半", 30);
        CHINESE_NUM_MAP = Collections.unmodifiableMap(m);
    }

    /** Day-of-week Chinese character to java.time.DayOfWeek mapping */
    private static final java.util.Map<String, Integer> DOW_MAP;
    static {
        java.util.Map<String, Integer> m = new java.util.HashMap<>();
        m.put("一", 1); m.put("二", 2); m.put("三", 3); m.put("四", 4);
        m.put("五", 5); m.put("六", 6); m.put("日", 7); m.put("天", 7);
        DOW_MAP = Collections.unmodifiableMap(m);
    }

    @PostConstruct
    public void init() {
        try {
            Class<?> normalizerClass = Class.forName("com.time.nlp.TimeNormalizer");
            timeNormalizer = normalizerClass.getConstructor().newInstance();
            parseMethod = normalizerClass.getMethod("parse", String.class, String.class);
            getTimeUnitMethod = normalizerClass.getMethod("getTimeUnit");

            Class<?> timeUnitClass = Class.forName("com.time.nlp.TimeUnit");
            getTimeNormMethod = timeUnitClass.getMethod("getTimeNorm");
            getIsAllDayTimeMethod = timeUnitClass.getMethod("getIsAllDayTime");
            getOriginMethod = timeUnitClass.getMethod("getOrigin");

            log.info("TimeNormalizer initialized successfully");
        } catch (Exception e) {
            log.warn("Failed to initialize TimeNormalizer, will use fallback parser only: {}",
                    e.getMessage());
            timeNormalizer = null;
        }
    }

    /**
     * Parse a Chinese time expression into a list of TimeParseResult.
     * The base time defaults to the current system time.
     *
     * @param text the input text containing Chinese time expressions
     * @return list of parsed time results, empty if no time expression found
     */
    public List<TimeParseResult> parse(String text) {
        return parse(text, LocalDateTime.now());
    }

    /**
     * Parse a Chinese time expression into a list of TimeParseResult,
     * using the given base time for relative expressions.
     *
     * @param text     the input text containing Chinese time expressions
     * @param baseTime the reference time for relative expressions
     * @return list of parsed time results, empty if no time expression found
     */
    public List<TimeParseResult> parse(String text, LocalDateTime baseTime) {
        if (text == null || text.isBlank()) {
            return Collections.emptyList();
        }

        String trimmed = text.trim();
        log.debug("Parsing time expression: '{}' with base time: {}", trimmed, baseTime);

        // Try Time-NLP library first
        List<TimeParseResult> results = parseWithTimeNLP(trimmed, baseTime);
        if (!results.isEmpty()) {
            return results;
        }

        // Fall back to regex-based parsing
        log.debug("Time-NLP returned no results, using fallback parser for: '{}'", trimmed);
        results = parseWithFallback(trimmed, baseTime);
        return results;
    }

    /**
     * Parse and return the first (most likely) time result.
     *
     * @param text the input text containing Chinese time expressions
     * @return the parsed time result, or a result with null dateTime if not found
     */
    public TimeParseResult parseFirst(String text) {
        return parseFirst(text, LocalDateTime.now());
    }

    /**
     * Parse and return the first (most likely) time result.
     *
     * @param text     the input text
     * @param baseTime the reference time
     * @return the parsed time result
     */
    public TimeParseResult parseFirst(String text, LocalDateTime baseTime) {
        List<TimeParseResult> results = parse(text, baseTime);
        return results.isEmpty() ? TimeParseResult.builder()
                .originalExpression(text)
                .confidence(0.0)
                .build() : results.get(0);
    }

    // ==================== Time-NLP Library Integration ====================

    /**
     * Parse using the Time-NLP library (via reflection).
     */
    private List<TimeParseResult> parseWithTimeNLP(String text, LocalDateTime baseTime) {
        if (timeNormalizer == null) {
            return Collections.emptyList();
        }

        try {
            String baseTimeStr = baseTime.format(NORM_FORMATTER);
            parseMethod.invoke(timeNormalizer, text, baseTimeStr);
            Object[] timeUnits = (Object[]) getTimeUnitMethod.invoke(timeNormalizer);

            if (timeUnits == null || timeUnits.length == 0) {
                return Collections.emptyList();
            }

            List<TimeParseResult> results = new ArrayList<>();
            for (Object unit : timeUnits) {
                try {
                    TimeParseResult result = convertTimeUnit(unit, text, baseTime);
                    if (result != null && result.isParsed()) {
                        results.add(result);
                    }
                } catch (Exception e) {
                    log.warn("Failed to convert TimeUnit for text '{}': {}", text, e.getMessage());
                }
            }

            return results;
        } catch (Exception e) {
            log.warn("Time-NLP parsing failed for '{}': {}", text, e.getMessage());
            return Collections.emptyList();
        }
    }

    /**
     * Convert a Time-NLP TimeUnit (accessed via reflection) into our TimeParseResult.
     */
    private TimeParseResult convertTimeUnit(Object unit, String originalText,
                                             LocalDateTime baseTime) throws Exception {
        String timeNorm = (String) getTimeNormMethod.invoke(unit);
        if (timeNorm == null || timeNorm.isBlank()) {
            return null;
        }

        boolean isAllDay = (boolean) getIsAllDayTimeMethod.invoke(unit);
        LocalDateTime dateTime = null;
        boolean ambiguous = false;
        String disambiguation = null;

        try {
            if (isAllDay) {
                LocalDate date = LocalDate.parse(timeNorm.trim(), NORM_DATE_FORMATTER);
                dateTime = date.atStartOfDay();
            } else {
                dateTime = LocalDateTime.parse(timeNorm.trim(), NORM_FORMATTER);
            }
        } catch (DateTimeParseException e) {
            log.warn("Failed to parse normalized time string '{}': {}", timeNorm, e.getMessage());
            dateTime = tryParseFlexible(timeNorm);
        }

        if (dateTime != null) {
            int hour = dateTime.getHour();
            if (originalText.contains("下午") || originalText.contains("晚上")
                    || originalText.contains("傍晚") || originalText.contains("晚间")) {
                if (hour < 12) {
                    dateTime = dateTime.plusHours(12);
                    ambiguous = true;
                    disambiguation = "Expression contains PM marker, adjusted from "
                            + hour + ":00 to " + dateTime.getHour() + ":00";
                }
            } else if (originalText.contains("凌晨") || originalText.contains("早上")
                    || originalText.contains("上午")) {
                if (hour >= 12 && hour != 12) {
                    ambiguous = true;
                    disambiguation = "Expression contains AM marker, resolved to "
                            + hour + ":00";
                }
            }
        }

        double confidence = ambiguous ? 0.75 : 0.85;

        String origin = (String) getOriginMethod.invoke(unit);
        return TimeParseResult.builder()
                .dateTime(dateTime)
                .originalExpression(origin != null ? origin : originalText)
                .normalizedTime(timeNorm)
                .allDay(isAllDay)
                .wasAmbiguous(ambiguous)
                .disambiguationNote(disambiguation)
                .confidence(confidence)
                .build();
    }

    /**
     * Try to parse a time string with flexible formats.
     */
    private LocalDateTime tryParseFlexible(String timeNorm) {
        if (timeNorm == null) return null;
        String trimmed = timeNorm.trim();

        // Try date-only
        try {
            return LocalDate.parse(trimmed, NORM_DATE_FORMATTER).atStartOfDay();
        } catch (DateTimeParseException ignored) {}

        // Try with different separators
        try {
            String normalized = trimmed.replace("/", "-");
            return LocalDateTime.parse(normalized, NORM_FORMATTER);
        } catch (DateTimeParseException ignored) {}

        return null;
    }

    // ==================== Fallback Regex-Based Parser ====================

    /**
     * Fallback parser using regex patterns for common Chinese time expressions.
     * Used when the Time-NLP library is unavailable or fails to parse.
     */
    private List<TimeParseResult> parseWithFallback(String text, LocalDateTime baseTime) {
        List<TimeParseResult> results = new ArrayList<>();

        LocalDate datePart = resolveDate(text, baseTime);
        LocalTime timePart = resolveTime(text);

        // No date or time pattern matched — not a time expression
        if (datePart == null && timePart == null) {
            return Collections.emptyList();
        }

        // If only time was found without a date, default to the base date
        if (datePart == null) {
            datePart = baseTime.toLocalDate();
        }
        boolean allDay = (timePart == null);
        boolean ambiguous = false;
        String disambiguation = null;

        if (timePart == null) {
            timePart = defaultTimeForPeriod(text);
        }

        // Disambiguation for PM markers with low hours
        if (timePart.getHour() < 12) {
            boolean hasPmMarker = text.contains("下午") || text.contains("晚上")
                    || text.contains("傍晚") || text.contains("晚间");
            boolean hasAmMarker = text.contains("凌晨") || text.contains("早上")
                    || text.contains("上午");

            if (hasPmMarker) {
                timePart = timePart.plusHours(12);
                ambiguous = true;
                disambiguation = "PM marker detected, adjusted to afternoon";
            } else if (!hasAmMarker && timePart.getHour() >= 1 && timePart.getHour() <= 6) {
                // Ambiguous: "两点" without context -- could be AM or PM
                // Default to PM for user convenience in calendar context
                ambiguous = true;
                int originalHour = timePart.getHour();
                timePart = timePart.plusHours(12);
                disambiguation = "No AM/PM marker; defaulted to afternoon ("
                        + timePart.getHour() + ":00). "
                        + "Original was " + originalHour + ":00";
            }
        }

        LocalDateTime dateTime = LocalDateTime.of(datePart, timePart);

        results.add(TimeParseResult.builder()
                .dateTime(dateTime)
                .originalExpression(text)
                .normalizedTime(dateTime.format(NORM_FORMATTER))
                .allDay(allDay)
                .wasAmbiguous(ambiguous)
                .disambiguationNote(disambiguation)
                .confidence(0.70)
                .build());

        return results;
    }

    /**
     * Resolve the date part from a Chinese time expression.
     * Returns {@code null} when no date pattern matches.
     */
    private LocalDate resolveDate(String text, LocalDateTime baseTime) {
        LocalDate base = baseTime.toLocalDate();

        // Absolute date: 2024年3月15日
        Matcher absMatcher = ABSOLUTE_DATE_PATTERN.matcher(text);
        if (absMatcher.find()) {
            int year = absMatcher.group(1) != null ? Integer.parseInt(absMatcher.group(1)) : base.getYear();
            int month = Integer.parseInt(absMatcher.group(2));
            int day = Integer.parseInt(absMatcher.group(3));
            if (year < 100) year += 2000;
            try {
                return LocalDate.of(year, month, day);
            } catch (Exception e) {
                log.warn("Invalid absolute date: {}-{}-{}", year, month, day);
            }
        }

        // Relative day references
        Matcher relMatcher = RELATIVE_DAY_PATTERN.matcher(text);
        if (relMatcher.find()) {
            String word = relMatcher.group(1);
            switch (word) {
                case "今天":
                case "今晚":
                    return base;
                case "明天":
                    return base.plusDays(1);
                case "后天":
                    return base.plusDays(2);
                case "大后天":
                    return base.plusDays(3);
                case "昨天":
                    return base.minusDays(1);
                case "前天":
                    return base.minusDays(2);
            }
        }

        // "N天后"
        Matcher daysLaterMatcher = DAYS_LATER_PATTERN.matcher(text);
        if (daysLaterMatcher.find()) {
            int days = parseChineseNumber(daysLaterMatcher.group(1));
            if (days > 0) {
                return base.plusDays(days);
            }
        }

        // Week references: 下周一, 上周三, 本周五
        Matcher weekMatcher = WEEK_REF_PATTERN.matcher(text);
        if (weekMatcher.find()) {
            String direction = weekMatcher.group(1);
            String dowChar = weekMatcher.group(2);
            int targetDow = DOW_MAP.getOrDefault(dowChar, 1);
            int currentDow = base.getDayOfWeek().getValue();

            int offset;
            switch (direction) {
                case "下":
                    // Always land in the next calendar week (Mon–Sun).
                    int daysToNextMonday = (8 - currentDow) % 7;
                    if (daysToNextMonday == 0) daysToNextMonday = 7;
                    offset = daysToNextMonday + (targetDow - 1);
                    break;
                case "上":
                    offset = targetDow - currentDow;
                    if (offset >= 0) offset -= 7;
                    break;
                case "本":
                default:
                    offset = targetDow - currentDow;
                    if (offset < 0) offset += 7;
                    break;
            }
            return base.plusDays(offset);
        }

        // No date pattern matched
        return null;
    }

    /**
     * Resolve the time part from a Chinese time expression.
     */
    private LocalTime resolveTime(String text) {
        // Check for hour expression
        Matcher hourMatcher = HOUR_PATTERN.matcher(text);
        if (hourMatcher.find()) {
            int hour = parseChineseNumber(hourMatcher.group(1));
            int minute = 0;

            if (hourMatcher.group(2) != null) {
                minute = parseChineseNumber(hourMatcher.group(2));
            }

            if (hour >= 0 && hour <= 23 && minute >= 0 && minute <= 59) {
                return LocalTime.of(hour, minute);
            }
        }

        // Check for "N个小时后" or "N分钟后"
        Matcher hoursLaterMatcher = HOURS_LATER_PATTERN.matcher(text);
        if (hoursLaterMatcher.find()) {
            int hours = Integer.parseInt(hoursLaterMatcher.group(1));
            return LocalTime.now().plusHours(hours);
        }

        Matcher minutesLaterMatcher = MINUTES_LATER_PATTERN.matcher(text);
        if (minutesLaterMatcher.find()) {
            int minutes = Integer.parseInt(minutesLaterMatcher.group(1));
            return LocalTime.now().plusMinutes(minutes);
        }

        return null;
    }

    /**
     * Return a sensible default time when the expression contains a time-of-day
     * marker (e.g. "晚上", "下午") but no explicit hour.
     */
    private LocalTime defaultTimeForPeriod(String text) {
        if (text.contains("凌晨"))   return LocalTime.of(6, 0);
        if (text.contains("早上"))   return LocalTime.of(8, 0);
        if (text.contains("上午"))   return LocalTime.of(9, 0);
        if (text.contains("中午"))   return LocalTime.of(12, 0);
        if (text.contains("下午"))   return LocalTime.of(14, 0);
        if (text.contains("傍晚"))   return LocalTime.of(18, 0);
        if (text.contains("晚上") || text.contains("晚间") || text.contains("今晚"))
            return LocalTime.of(19, 0);
        return LocalTime.MIDNIGHT;
    }

    /**
     * Parse a Chinese or Arabic numeral string into an integer.
     */
    private int parseChineseNumber(String numStr) {
        if (numStr == null || numStr.isBlank()) return 0;

        // Try Arabic numeral first
        try {
            return Integer.parseInt(numStr.trim());
        } catch (NumberFormatException ignored) {}

        // Look up in Chinese numeral map
        Integer mapped = CHINESE_NUM_MAP.get(numStr.trim());
        if (mapped != null) return mapped;

        // Handle compound Chinese numbers like "十一点" -> parse "十一"
        String trimmed = numStr.trim();
        if (trimmed.startsWith("十")) {
            Integer tens = CHINESE_NUM_MAP.get("十");
            if (trimmed.length() == 1) return tens;
            String rest = trimmed.substring(1);
            Integer ones = CHINESE_NUM_MAP.get(rest);
            if (ones != null) return tens + ones;
        }

        log.warn("Could not parse Chinese number: '{}'", numStr);
        return 0;
    }
}

File: voicecal-mobile/src/services/nlu/TimeParserService.ts
import dayjs from 'dayjs';

// Chinese numeral mapping
const CHINESE_NUM_MAP: Record<string, number> = {
  '零': 0, '一': 1, '二': 2, '两': 2,
  '三': 3, '四': 4, '五': 5, '六': 6,
  '七': 7, '八': 8, '九': 9, '十': 10,
  '十一': 11, '十二': 12, '十三': 13, '十四': 14, '十五': 15,
  '十六': 16, '十七': 17, '十八': 18, '十九': 19,
  '二十': 20, '二十一': 21, '二十二': 22, '二十三': 23, '二十四': 24,
  '半': 30,
};

// Day of week mapping (Monday=1 ... Sunday=7)
const DOW_MAP: Record<string, number> = {
  '一': 1, '二': 2, '三': 3, '四': 4, '五': 5, '六': 6, '日': 7, '天': 7,
};

// Default time for time-of-day periods
const DEFAULT_TIME_FOR_PERIOD: Record<string, string> = {
  '凌晨': '06:00',
  '早上': '08:00',
  '上午': '09:00',
  '中午': '12:00',
  '下午': '14:00',
  '傍晚': '18:00',
  '晚上': '19:00',
  '晚间': '19:00',
  '今晚': '19:00',
};

export interface TimeParseResult {
  /** ISO datetime string (YYYY-MM-DDTHH:mm:ss) or null if unparseable */
  dateTime: string | null;
  /** Date portion as YYYY-MM-DD */
  date: string | null;
  /** Time portion as HH:mm */
  time: string | null;
  /** The original input text that was parsed */
  originalExpression: string;
  /** Whether AM/PM disambiguation was applied */
  wasAmbiguous: boolean;
  /** Explanation of any disambiguation that was applied */
  disambiguationNote?: string;
  /** Confidence score (0.0 - 1.0) */
  confidence: number;
}

// ========== Regex patterns (ported from Java, with Unicode flag for JS) ==========

/** Absolute date: 2024年3月15日, 3月15号 */
const ABSOLUTE_DATE_PATTERN = /(?:(\d{2,4})\s*年)?(\d{1,2})\s*月(\d{1,2})\s*[日号]/u;

/** Relative day references */
const RELATIVE_DAY_PATTERN = /(大后天|后天|明天|昨天|前天|今天|今晚)/u;

/** "N天后" / "N天以后" (Arabic or Chinese numerals) */
const DAYS_LATER_PATTERN = /([一二两三四五六七八九十\d]+)\s*天(?:以后|后)/u;

/** "N个小时后" / "N小时后" */
const HOURS_LATER_PATTERN = /(\d+)\s*(?:个)?小时(?:以后|后)/u;

/** "N分钟后" / "N分钟以后" */
const MINUTES_LATER_PATTERN = /(\d+)\s*分钟(?:以后|后)/u;

/** Week references: 下周一, 上周三, 本周五 */
const WEEK_REF_PATTERN = /(上|下|本)(?:周|星期|礼拜)([一二三四五六日天])/u;

/** Explicit hour: "两点", "十点", "3点", "两点半", "十点十五分" */
const HOUR_PATTERN = /(\d|[一二两三四五六七八九十]+)\s*点(?:(半|[一二三四五六七八九十\d]+)\s*分?)?/u;

// ========== Service ==========

/**
 * Service for parsing Chinese natural language time expressions into
 * structured date/time values.  Ported from the Java TimeParserService
 * fallback regex-based parser.
 */
export const TimeParserService = {
  /**
   * Parse a Chinese or Arabic numeral string into an integer.
   *
   * Handles:
   * - Arabic digits via parseInt
   * - Single Chinese chars from CHINESE_NUM_MAP
   * - Compound forms: "十一" (11), "二十" (20), "二十一" (21)
   * - "半" returns 30
   */
  parseChineseNumber(str: string): number {
    if (!str || !str.trim()) return 0;

    const trimmed = str.trim();

    // Try Arabic numeral first
    const asInt = parseInt(trimmed, 10);
    if (!isNaN(asInt)) return asInt;

    // Direct lookup in the map (covers single chars, compound entries, and "半")
    if (CHINESE_NUM_MAP[trimmed] !== undefined) {
      return CHINESE_NUM_MAP[trimmed];
    }

    // Handle compound Chinese numbers starting with "十"
    // e.g. "十" -> 10, "十一" -> 11, "十二" -> 12 ...
    if (trimmed.startsWith('十')) {
      const tens = CHINESE_NUM_MAP['十']; // 10
      if (trimmed.length === 1) return tens;
      const rest = trimmed.substring(1);
      const ones = CHINESE_NUM_MAP[rest];
      if (ones !== undefined) return tens + ones;
    }

    return 0;
  },

  /**
   * Resolve the date part from a Chinese time expression.
   * Returns a YYYY-MM-DD string or null when no date pattern matches.
   *
   * @param text     the input text
   * @param baseDate reference date for relative expressions (YYYY-MM-DD)
   */
  resolveDate(text: string, baseDate: string): string | null {
    const base = dayjs(baseDate);

    // 1. Absolute date: 2024年3月15日 / 3月15号
    const absMatch = ABSOLUTE_DATE_PATTERN.exec(text);
    if (absMatch) {
      let year = absMatch[1] != null ? parseInt(absMatch[1], 10) : base.year();
      const month = parseInt(absMatch[2], 10);
      const day = parseInt(absMatch[3], 10);
      if (year < 100) year += 2000;
      try {
        const candidate = dayjs(`${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`);
        if (candidate.isValid()) return candidate.format('YYYY-MM-DD');
      } catch {
        // invalid date, fall through
      }
    }

    // 2. Relative day references
    const relMatch = RELATIVE_DAY_PATTERN.exec(text);
    if (relMatch) {
      const word = relMatch[1];
      switch (word) {
        case '今天':
        case '今晚':
          return base.format('YYYY-MM-DD');
        case '明天':
          return base.add(1, 'day').format('YYYY-MM-DD');
        case '后天':
          return base.add(2, 'day').format('YYYY-MM-DD');
        case '大后天':
          return base.add(3, 'day').format('YYYY-MM-DD');
        case '昨天':
          return base.subtract(1, 'day').format('YYYY-MM-DD');
        case '前天':
          return base.subtract(2, 'day').format('YYYY-MM-DD');
      }
    }

    // 3. "N天后" / "N天以后"
    const daysLaterMatch = DAYS_LATER_PATTERN.exec(text);
    if (daysLaterMatch) {
      const days = this.parseChineseNumber(daysLaterMatch[1]);
      if (days > 0) {
        return base.add(days, 'day').format('YYYY-MM-DD');
      }
    }

    // 4. Week references: 下周一, 上周三, 本周五
    const weekMatch = WEEK_REF_PATTERN.exec(text);
    if (weekMatch) {
      const direction = weekMatch[1];
      const dowChar = weekMatch[2];
      const targetDow = DOW_MAP[dowChar] ?? 1;
      const currentDow = base.day() === 0 ? 7 : base.day(); // dayjs: 0=Sun

      let offset: number;
      switch (direction) {
        case '下': {
          // Always land in the next calendar week (Mon-Sun).
          let daysToNextMonday = (8 - currentDow) % 7;
          if (daysToNextMonday === 0) daysToNextMonday = 7;
          offset = daysToNextMonday + (targetDow - 1);
          break;
        }
        case '上': {
          offset = targetDow - currentDow;
          if (offset >= 0) offset -= 7;
          break;
        }
        case '本':
        default: {
          offset = targetDow - currentDow;
          if (offset < 0) offset += 7;
          break;
        }
      }
      return base.add(offset, 'day').format('YYYY-MM-DD');
    }

    // No date pattern matched
    return null;
  },

  /**
   * Resolve the time part from a Chinese time expression.
   * Returns HH:mm or null.
   *
   * @param text the input text
   */
  resolveTime(text: string): string | null {
    // 1. Explicit hour expression: "两点", "十点半", "3点十五分"
    const hourMatch = HOUR_PATTERN.exec(text);
    if (hourMatch) {
      const hour = this.parseChineseNumber(hourMatch[1]);
      let minute = 0;

      if (hourMatch[2] != null) {
        minute = this.parseChineseNumber(hourMatch[2]);
      }

      if (hour >= 0 && hour <= 23 && minute >= 0 && minute <= 59) {
        return `${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}`;
      }
    }

    // 2. "N个小时后" / "N小时后"
    const hoursLaterMatch = HOURS_LATER_PATTERN.exec(text);
    if (hoursLaterMatch) {
      const hours = parseInt(hoursLaterMatch[1], 10);
      return dayjs().add(hours, 'hour').format('HH:mm');
    }

    // 3. "N分钟后"
    const minutesLaterMatch = MINUTES_LATER_PATTERN.exec(text);
    if (minutesLaterMatch) {
      const minutes = parseInt(minutesLaterMatch[1], 10);
      return dayjs().add(minutes, 'minute').format('HH:mm');
    }

    return null;
  },

  /**
   * Check if text contains any time-of-day keywords and return the default
   * time for that period. Returns '00:00' if no period keyword is found.
   *
   * @param text the input text
   */
  defaultTimeForPeriod(text: string): string {
    // Check each period keyword in order; first match wins
    for (const [keyword, time] of Object.entries(DEFAULT_TIME_FOR_PERIOD)) {
      if (text.includes(keyword)) {
        return time;
      }
    }
    return '00:00';
  },

  /**
   * AM/PM disambiguation for parsed hour values.
   *
   * - If hour < 12 and text contains PM markers ("下午", "晚上", "傍晚", "晚间"),
   *   add 12 to the hour.
   * - If no marker at all and hour is 1-6, default to PM (add 12) for user
   *   convenience in a calendar context.
   *
   * @param hour  the parsed hour (0-23)
   * @param minute the parsed minute (kept as-is, included in return for convenience)
   * @param text  the original input text
   */
  disambiguateTime(
    hour: number,
    minute: number,
    text: string,
  ): { hour: number; wasAmbiguous: boolean; note?: string } {
    const hasPmMarker =
      text.includes('下午') ||
      text.includes('晚上') ||
      text.includes('傍晚') ||
      text.includes('晚间');

    const hasAmMarker =
      text.includes('凌晨') ||
      text.includes('早上') ||
      text.includes('上午');

    if (hour < 12) {
      if (hasPmMarker) {
        return {
          hour: hour + 12,
          wasAmbiguous: true,
          note: '检测到下午/晚上标记，已调整为下午时间',
        };
      }

      if (!hasAmMarker && hour >= 1 && hour <= 6) {
        // Ambiguous: "两点" without context -- could be AM or PM.
        // Default to PM for user convenience in calendar context.
        return {
          hour: hour + 12,
          wasAmbiguous: true,
          note: `未检测到上午/下午标记，默认为下午（${hour + 12}:00），原始时间为${hour}:00`,
        };
      }
    }

    return { hour, wasAmbiguous: false };
  },

  /**
   * Main entry point. Parse a Chinese time expression and return a structured
   * result with date, time, confidence, and disambiguation info.
   *
   * @param text      the input text containing Chinese time expressions
   * @param baseDate  optional reference date as YYYY-MM-DD (defaults to today)
   */
  parse(text: string, baseDate?: string): TimeParseResult {
    const base = baseDate ?? dayjs().format('YYYY-MM-DD');

    if (!text || !text.trim()) {
      return {
        dateTime: null,
        date: null,
        time: null,
        originalExpression: text ?? '',
        wasAmbiguous: false,
        confidence: 0,
      };
    }

    const trimmed = text.trim();

    // Resolve date and time independently
    let datePart = this.resolveDate(trimmed, base);
    let timePart = this.resolveTime(trimmed);

    // No date or time pattern matched -- not a time expression
    if (datePart === null && timePart === null) {
      return {
        dateTime: null,
        date: null,
        time: null,
        originalExpression: trimmed,
        wasAmbiguous: false,
        confidence: 0,
      };
    }

    // If only time was found without a date, default to the base date
    const allDay = timePart === null;
    if (datePart === null) {
      datePart = base;
    }

    let ambiguous = false;
    let disambiguation: string | undefined;

    // If no explicit time was found, supply a default based on period keywords
    if (timePart === null) {
      timePart = this.defaultTimeForPeriod(trimmed);
    }

    // Apply AM/PM disambiguation
    const [hourStr, minuteStr] = timePart.split(':');
    let hour = parseInt(hourStr, 10);
    const minute = parseInt(minuteStr, 10);

    if (!allDay) {
      const result = this.disambiguateTime(hour, minute, trimmed);
      hour = result.hour;
      ambiguous = result.wasAmbiguous;
      disambiguation = result.note;
    }

    const finalTime = `${String(hour).padStart(2, '0')}:${String(minute).padStart(2, '0')}`;
    const dateTime = `${datePart}T${finalTime}:00`;

    return {
      dateTime,
      date: datePart,
      time: finalTime,
      originalExpression: trimmed,
      wasAmbiguous: ambiguous,
      disambiguationNote: disambiguation,
      confidence: 0.70,
    };
  },
};
